### PR TITLE
ENH: Add support for engine=numba to groupby apply

### LINKED ISSUE #53933: API/DISC: Mock pandas DataFrame/Series/Index in numba agg/apply/transform
## Overview

I was working on numba apply support pretty recently, and I was finding it pretty hard to support apply
the same way as agg and transform(which is to take in ``values`` and ``index`` and return the new ``values``)
because it's way more flexible than agg or transform 
(You can change the index as well as the values, and also change shape of values by adding/deleting columns).

This made me end up writing a ton of code to handle this (basically reimplementing the concat logic for apply for groupby), so I decided to take a look at wrapping the DataFrame object using the numba extension API.

## Pros
- Same API between numba and regular engine, you can reuse functions since you would use the DF object for both
   - Small nit: We wouldn't support stuff that numba doesn't support in nopython mode like non-string names
   - It's also more convenient as a user to not have to keep track the index/columns/values of the DataFrame separately.
- Reuse internal logic (what goes in out of the numba function is a DF/Series because we do the boxing/unboxing)
   - Should be able to de-dup a lot, and save a lot of misery in keeping the numba/regular paths consistent

## Cons
- Annoying to develop
  - Hard to wrap classes
     - You are basically writing C code in Python via numba that tells numba how to box/unbox the Python DF to a C representation
  - Adding new methods is pretty easy tho, those are just like regular numba functions
  - Easy to make mistakes

We should decide whether we want to go down this road, and if so, how much of the pandas API to implement.
(e.g. what methods should exist, and what to wrap)

Some sample code (only works with int64 arrays ATM, which is hardcoded) wrapping a subset of DataFrame attributes/methods.

cc @mroeschke @jbrockmendel @rhshadrach

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #53933  (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

I ended up just wrapping the DF/Series objects so they're recognized by numba.

Currently, all you can do with the DF/Series objects is extract the values and the index from them.

I think it would make sense to implement things like the addition operators, and some of the common reductions on them, but that's for the future.

I went this way since going through the values, index way of the other functions was just way to much of a hassle, and would cause a lot of code duplication and re-implementation. More details in #53933 
This way also lets users write a single function for the "python"/"numba" engines.

Currently done:
- Reductions (to scalar, axis 0, axis 1)
- Transforms ( tested with normalization func, and identity func)
- Expansions (the example in the user guide is tested)

TODO:
- parallel support
   - numba containers are not thread safe. Will look into threading manually with Python and hope the GIL doesn't get in the way (for a follow up).
- Series support in groupby

### CODE REVIEW

**Comment by lithomas1** [comment]
OK, this should be ready for review now.

As of right now, this PR does 2 things

1. Allows using pandas Index/Series/DataFrame within numba jitted code.
   a. I can split this part into another PR which should reduce the diff down to 100-200 lines.
2. Allows passing a function to jit with numba into groupby.apply

The API for this is currently really clunky since as of right now, a lot of the pandas API is not implemented yet.
(Basically only the constructors are implemented).

(You can see how you would write a function for the numba engine in ``pandas/tests/numba/test_basic.py``, and ``pandas/tests/groupby/test_numba.py``)

However, all the basic operations I can think of (reductions, transformations, adding a column) should work right now, so it's at least usable in the current state.

And, looking on the bright side, it can only get better from here :) .

Right now, I'm not really looking for a deep review, I'm just trying to get confirmation that this is the right approach.
(I'll clean everything up after that).

File: pandas/core/_numba/extensions.py [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+        layout = layout or self.layout
+        return type(self)(dtype, layout, self.pyclass)
+
+
+class SeriesType(types.ArrayCompatible):
+    """
+    The type class for Series objects.
+    """
+
+    array_priority = 1000
```
Comment (jbrockmendel) [inline pandas/core/_numba/extensions.py:67]: this is distinct from the dunder `__array_priority__`?

File: pandas/core/_numba/extensions.py [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+
+    def __init__(self, dtype, index) -> None:
+        assert isinstance(index, IndexType)
+        self.dtype = dtype
+        self.index = index
+        self.values = types.Array(self.dtype, 1, "C")
```
Comment (jbrockmendel) [inline pandas/core/_numba/extensions.py:73]: can this be something more distinctive than "values"?  we use that to mean a bunch of different things around the codebase

File: pandas/core/_numba/extensions.py [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+class DataFrameType(types.ArrayCompatible):
+    """
+    The type class for DataFrame objects.
+    """
+
+    array_priority = 1000
+
+    def __init__(self, dtype, index, layout, columns) -> None:
+        assert isinstance(index, IndexType)
+        self.dtype = dtype
```
Comment (jbrockmendel) [inline pandas/core/_numba/extensions.py:102]: annotation?  i dont know what to expect here

**Comment by jbrockmendel** [comment]
Looked through it and made some superficial comments, don't really have much to add.  I'm not a huge fan of the engine/engine_kwargs pattern (xref #53239) so am not enthusiastic about this, but not opposed.  

Does this make any implicit assumptions about having all-numpy dtypes?

Would there be a viable path for this to live in numba or numba-foo (and not be our problem)?

**Comment by lithomas1** [comment]
Thanks for the review.

> Looked through it and made some superficial comments, don't really have much to add. I'm not a huge fan of the engine/engine_kwargs pattern (xref #53239) so am not enthusiastic about this, but not opposed.
> 
> Does this make any implicit assumptions about having all-numpy dtypes?

This assumes all numpy for now. I don't think it's too hard to implement this for Arrow, you just need to know the Arrow columnar format well.

(e.g. see https://uwekorn.com/2018/08/03/use-numba-to-work-with-apache-arrow-in-pure-python.html)

> Would there be a viable path for this to live in numba or numba-foo (and not be our problem)?

The way extensions are registered for numba(they use the entrypoints mechanism like we do for plotting), we can definitely move ``pandas/core/_numba/extensions.py`` outside to a separate repo.

Two things we'd need to figure out if we go down this road are:

1) Using pandas internals (Cython code).
    Numba provides a mechanism to call Cython functions from Python code, which could be useful for implementing indexing.

2) Relationship between pandas and this library

     We'd also need to take a optional dependency on the split off library or provide some sort of plugin structure for things 
     such as ``Groupby.apply``, since the looping part is not the same for the Python/numba engines.

I'll add something to the agenda for the dev meeting on Wednesday.

**Comment by jbrockmendel** [comment]
> We'd also need to take a optional dependency on the split off library or provide some sort of plugin structure for things
such as Groupby.apply, since the looping part is not the same for the Python/numba engines.

Unless we expect more of these to pop up, then a plugin structure sounds less appealing than having this live inside pandas.  So pls ignore the previous question about having this live elsewhere.

File: pandas/core/_numba/extensions.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+
+    def copy(self, dtype=None, ndim: int = 1, layout: str = "C"):
+        assert ndim == 1
+        assert layout == "C"
+        if dtype is None:
+            dtype = self.dtype
+        return type(self)(dtype, self.index)
+
+
+class DataFrameType(types.ArrayCompatible):
```
Comment (mroeschke) [inline pandas/core/_numba/extensions.py:93]: Based on https://numba.discourse.group/t/implementing-pandas-dataframe-type-via-numba-extension-types/2005/2, could we borrow what cudf has done?

**Pull request closed without merging**